### ISSUE TITLE
Question about the network

### ISSUE THREAD

**Issue by GitStarter** [JiahuiYu/generative_inpainting#375]
Hello Jiahui, thanks for sharing such wonderful work. I read through the paper and have some questions regards to both structure and implementation. Sorry I am new to machine learning and my questions may seems a bit naive.

Q1. why did you use ELU for activation of the x and sigmoid activation for the mask? are these just on experimental trial? I searched around and most of the articles suggest to avoid use the sigmoid activation it kills the gradient in deep neural network. the ELU does not have the gradient kill issue but it is very computational heavy compare to ReLU or leaky ReLU.

Q2. the cnum is number of filters used right? is the number already reduced by 25% as indicated in the paper? like 48 = 64*0.75? i saw some post uses less filter numbers. is it going to affect the training result significantly? and what would be the suggested range to pick that number from?

Q3. what is are those parameters under #train? 
train_spe: 4000
max_iters: 100000000
viz_max_out: 10
val_psteps: 2000
those are the default value i got and i am not sure what effect they have on the training process.

Q4, I am trying to train the ImageNet and for my local PC, i have Nvidia GTX 1060 6GB but it cannot go up to 16 batch_size, and i tried to train them with around 8 batch size but the result shows blurry inpainted region. should I adjust the l1_loss_alpha to the 0.5? or 2? I saw two issue post suggesting different values. and I only have these three parameters under #loss section.
ae_loss: True
l1_loss: True
l1_loss_alpha: 1.

Q5. Since I don't have enough graphic memories, I applied for google cloud compute engine and have VM created using their deep learning VM from the market place. with same version of the tensorflow and tensorflow-gpu from my local computer, the VM was not able to recognize the GPU installed. I tried change the trainer to multiGPU trainer and setup the parameters in inpaint.yml. I am only using one GPU for the VM and I was using the tesla V100 16gb and the network gpu utilization just blows up and the kernel got stuck and died. in the live nvidia-smi gpu monitoring I saw error like "cannot allocate the memory". after it blows up.

Thank you very much.

**Comment by JiahuiYu** [comment]
Q1: ELU works better empirically. But you can definitely try to use ReLU and compare their performance. I don't see any gradient vanishing problem in our setting.

Q2: I don't have any related experiments when filter numbers are reduced further. I guess the performance will be worse.

Q3: Those parameters do not affect the training. Those parameters are documented in the code if you want to learn their usages.

Q4: In general batch size less than 16 will lead to worse performance. I'd suggest to reduce image resolution in your case.

Q5: #92 #68 These issues may help. But first you will need to make sure that your installed machines are correctly configured to run tensorflow code.

If you have more questions, you can leave them here.

**Comment by GitStarter** [comment]
> Q3: Those parameters do not affect the training. Those parameters are documented in the code if you want to learn their usages.
> 
> Q4: In general batch size less than 16 will lead to worse performance. I'd suggest to reduce image resolution in your case.

Thank you very much for your response, 
for Q3, could you please link the file where it is documented? in the inpaint.yml I have, I only see the documentation for the 

Q4. when reducing the image resolution, if I reduce the shape to [128,128,3] I shall also reduce the 
height: 128
width: 128
max_delta_height: 32
max_delta_width: 32
to half of their value right? since they define the parameters used to generate the mask?

Also, if the result is blurred region, I shall increase the l1_loss_alpha  to 2 and tried it out?

Another question I have is that since ImageNet contains millions of images, I only picked the first 150 picture [0:150] for the 1000 classes for training images; and 10 pictures after that from 1000 classes [150:160] for validation images (both from the training image folder of the ImageNet). This setup should have 150k images for training and 10k for validation. this should be enough sample to train the network?

**Comment by JiahuiYu** [comment]
Q3: e.g., https://github.com/JiahuiYu/generative_inpainting/blob/master/train.py#L98

Q4: Yes.

Another Question: I don't know and you should have a try.

**Comment by GitStarter** [comment]
Thank you,

I have reconfigured another VM and it ran smoothly, right now I am using V100 to train half of the ImageNet dataset with a batch size of 24. and tried out leaky.relu for activation. it just finished one epoch. I am wondering how to calculate the time and speed? it says 1.23 batches/sec. 
Q5. In train_shuffled.flist i have 640610 image paths. I am wondering if the train_spe I shall increase to a large number instead of using 4000? I have zero experience in training the GAN so i don't know what would be a good number to start with.

Q6. from #231  it states that the train_spe is the number of iteration per epoch, does that means each epoch will have 4000 batches used for passing the loop, each batch has batch_size of 24? so training for one epoch will be 4000*1.23 = 4920s in my case above? and it will run in total max_iter then stop? which means it will continue train (max_iter/train_spe) number of epochs or we interrupted it?

**Comment by JiahuiYu** [comment]
@GitStarter 
Q5: You don't need to increase train_spe, as in each epoch, it will shuffle the file list.

Q6: Yes.

**Comment by GitStarter** [comment]
Thank you,

Q7. To use guided training, I need to prepare the _edge images. for example, we can run HED on the training images to create black/white [255/0] masks where black represents edge detected for the whole image. and the network will only train the mask region which is randomly generated during training?

Q8.Do I need to preprocess the training images to create the edge images and create a different flist file? How the flist file should be prepared?
```
    if FLAGS.guided:
        fnames = [(fname, fname[:-4] + '_edge.jpg') for fname in fnames]
        img_shapes = [img_shapes, img_shapes]
```
based on the code you gave, I believe I just ran the edge detector through the training data. and store the edge pictures in the same folder since their 'fname' part is the same? for example, in the same folder, I should have './train/folder1/image1.jpeg' and corresponding edge picture with name './train/folder1/image1_edge.jpg'?

Q9, the edge should also be three channels with the same value, (the same shape as the image), not 1 channel black and white picture correct?

**Comment by GitStarter** [comment]
![1](https://user-images.githubusercontent.com/7672367/70685943-ddbc9b80-1c78-11ea-8f94-5a33f9d49dcf.png)
![2](https://user-images.githubusercontent.com/7672367/70685944-ddbc9b80-1c78-11ea-9c2f-1cb8d0e4ea1e.png)
![3](https://user-images.githubusercontent.com/7672367/70685945-ddbc9b80-1c78-11ea-8e2d-a1c7719e94f2.png)
![4](https://user-images.githubusercontent.com/7672367/70685946-ddbc9b80-1c78-11ea-83cc-05ad729beb3c.png)
![5](https://user-images.githubusercontent.com/7672367/70685947-ddbc9b80-1c78-11ea-8be6-cac917bff3ca.png)
![6](https://user-images.githubusercontent.com/7672367/70685948-de553200-1c78-11ea-9ed7-7a3cd803792f.png)
![7](https://user-images.githubusercontent.com/7672367/70685949-de553200-1c78-11ea-9afb-6efc9ffa6d2b.png)
Here is the result of the eighth epoch, I used pre-trained model and trained with 24 batch_size. I am wonder does those blur just because it is at the beginning of the training or the pixel-wise loss is too low?

**Comment by JiahuiYu** [comment]
No. It is because more training is required.

**Comment by GitStarter** [comment]
Thank you very much for your patient.

I am wondering what kind of edge image should be prepared for. After I ran the released HED with their trained model got output of single channel 0-255 gray scale. and when I saved them to jpg image, it became three channel and all the 3 channel has the same value. 

as you mentioned in the other issue #99 threshold is 0.6.  does the released code will read and change the type to float then apply threshold?

**Comment by JiahuiYu** [comment]
This code will not produce HED edges. Please pre-processing your images with official HED code.

**Comment by GitStarter** [comment]
Yes, I am using the HED try to pre-process the HED image.
![cat](https://user-images.githubusercontent.com/7672367/70750620-bb6c6180-1cfc-11ea-8626-3a370c0db68e.jpg)
![cHED (1)](https://user-images.githubusercontent.com/7672367/70750621-bb6c6180-1cfc-11ea-9bb2-6cd0fd541776.jpg)
For example. i processed a 'cat.jpg'. so the HED edge image should be saved as 'cat_edge.jpg'. and put in the same folder right?

these are all steps required for preparing the guided images?

**Comment by Iris6713** [comment]
@GitStarter Does your finetune work well？
I have trained 80 epochs, but the effect is still blurry
How did you do it?

**Comment by GitStarter** [comment]
@Iris6713 I am running at 47 epochs right now and still have blurry effect. for imagenet I believe it requires some extra help with guided edges. However, I am having trouble with automating the HED image generation with the dataset I have right now.
